README.md
`Proc::tored` is a perl module to make it simpler to manage a service using a pid file.

    use Proc::tored;

    my $proctor = Proc::tored->new(dir => '/tmp', name => 'my-service');

    # Call do_stuff while the service is running or until do_stuff returns false
    $proctor->service(\&do_stuff)
      or die sprintf('process %d is already running this service!', $proctor->running_pid);

    # Signal another process running this service to quit gracefully, throwing an
    # error if it does not self-terminate after 15 seconds.
    if (my $pid = $proctor->stop_running_process(15)) {
      die "process $pid is being stubborn!";
    }


Changes
{{$NEXT}}

0.20 2017-10-27
- Add Dist::Zilla::Plugin::MetaJSON (thanks manwar)

0.19 2017-08-15
- Remove broken test from t/Proc/tored/PidFile.t

0.18 2017-07-28
- Lock file logic moved to Proc::tored::LockFile
- Fixed warnings triggered when cleaning up lock/pid files whose parent
  directory has been removed
- Fixed bug permitting a newly started service to enter the paused state
  another instance is already in the locked state, causing services to stack
  while paused

0.17 2017-03-18
- Add newline to warning emitted when a signal is caught
- Regularly touch the pid file while running

0.16 2017-03-09
- Fixed bug allowing pause flag to have precedence over stop flag
- Emit a warning when a posix signal is received

0.15 2017-03-08
- Fixed bug allowing a guard object to be held inappropriately
- Fixed bug causing pid files to be overwritten/unlinked by forked child
  processes

0.14 2017-03-08
- Update to Auto::Mata 0.07
- Proc::tored::Machine is now a Moo class
- Proc::tored::PidFile is now a Moo class
- Refactored and tightened state machine workflow

0.13 2017-03-06
- Use Auto::Mata to track service state

0.12 2017-03-02
- On service shutdown, truncate pid file before removing in case there are open
  handles in other processes

0.11 2017-02-23
- Fixed bug in definition of SignalList type
- Fixed test failures on MSWin32
- Added test for Proc::tored::Types

0.10 2017-02-22
- Remove pidfile after service completes
- Configurable signal trapping

0.09 2017-02-21
- Added Proc::tored::Manager->stop_file and ->pause_file to override touch file
  paths

0.08 2017-02-20
- Service state is now persistent
- New feature: pause/resume

0.07 2017-02-18
- Fixed serious bug causing pid file to be removed when a forked process exits

0.06 2017-02-17
- Use a touch file to signal self-termination

0.05 2017-02-17
- Refactor locking to support more platforms
- Work around for MSWin32/threaded signal handling in unit tests

0.04 2017-02-16
- Use Path::Tiny->tempdir for pidfile directory in tests
- Fixed untrapped error when lockfile creation fails

0.03 2017-02-15
- Replace poll_wait_time attribute with parameter to
  Proc::tored::Role::Running->stop_running_process
- Use secondary lock file to avoid inconsistent flocking on mswin32

0.02 2017-02-14
- New interface
- Trap normal signals (TERM, INT, HUP, PIPE)
- Rewrote cleanup code to handle unexpected errors more reliably
- Fixed some inconsistent results in unit tests (hopefully)

0.01 2017-02-13
- First version


lib/Proc/tored/Flag.pm
package Proc::tored::Flag;
# ABSTRACT: Ties a runtime flag to the existence of a touch file

=head1 SYNOPSIS

  use Proc::tored::Flag;

  my $fnord = Proc::tored::Flag->new(touch_file_path => '/my/service/path');

  $fnord->set; # touch file created if not already there
  $fnord->is_set; # true

  $fnord->unset; # touch file removed if it exists
  $fnord->is_set; # false

  if ($fnord->is_set) {
    warn "forgot what to do";
    exit 1;
  }

=cut

use strict;
use warnings;
use Moo;
use Path::Tiny 'path';
use Types::Standard -types;

has touch_file_path => (
  is  => 'ro',
  isa => Str,
  required => 1,
);

has file => (
  is  => 'lazy',
  isa => InstanceOf['Path::Tiny'],
  handles => {
    set    => 'touch',
    unset  => 'remove',
    is_set => 'exists',
  },
);

sub _build_file { path($_[0]->touch_file_path) }

1;


lib/Proc/tored/Machine.pm
package Proc::tored::Machine;

use strict;
use warnings;
use Moo;
use Carp;
use Auto::Mata '!with';
use Proc::tored::Flag;
use Proc::tored::PidFile;
use Proc::tored::Types -types;
use Time::HiRes;
use Type::Utils qw(declare as where);
use Types::Standard -types;

use constant READY  => 'READY';
use constant STATUS => 'STATUS';
use constant LOCK   => 'LOCK';
use constant TOUCH  => 'TOUCH';
use constant STOP   => 'STOP';
use constant TERM   => 'TERM';

my $Lock    = Maybe[InstanceOf['Guard']];
my $Flag    = InstanceOf['Proc::tored::Flag'];
my $PidFile = InstanceOf['Proc::tored::PidFile'];

my $Proctor = declare 'Proctor', as Dict[
  pidfile => $PidFile,   # PidFile
  stopped => $Flag,      # Stop Flag
  paused  => $Flag,      # Pause Flag
  traps   => SignalList, # list of signals to trap
  call    => CodeRef,    # code ref to call while running
  lock    => $Lock,      # guard set on successful lock
  started => Bool,       # initialization status
  quit    => Bool,       # flag set by trapped posix signals
  finish  => Bool,       # true when last callback returned false
];

my $Stopped    = declare 'Stopped',    as  $Proctor, where { $_->{stopped}->is_set || $_->{quit} };
my $NotStopped = declare 'NotStopped', as ~$Stopped;
my $Unlocked   = declare 'Unlocked',   as  $NotStopped, where { !$_->{lock} };
my $Locked     = declare 'Locked',     as  $NotStopped, where { $_->{lock} };
my $Started    = declare 'Started',    as  $Locked, where { $_->{started} };
my $Paused     = declare 'Paused',     as  $Started, where { $_->{paused}->is_set };
my $NotPaused  = declare 'NotPaused',  as ~$Paused;
my $Running    = declare 'Running',    as  $NotPaused & $Started, where { !$_->{finish} };
my $Finished   = declare 'Finished',   as  $Started, where { $_->{finish} };

sub pause_sleep {
  my ($acc, $time) = @_;
  Time::HiRes::sleep($time);
  $acc;
}

sub sigtrap {
  my $acc = shift;
  foreach my $signal (@{$acc->{traps}}) {
    $SIG{$signal} = sub {
      warn "Caught SIG$signal\n";
      $acc->{quit} = 1;
      $acc;
    };
  }
}

my $FSM = machine {
  ready READY;
  term  TERM;

  # Ready
  transition READY,  to STATUS, on $Proctor;

  # Service loop
  transition STATUS, to STOP,   on $Finished;
  transition STATUS, to STOP,   on $Stopped;
  transition STATUS, to TOUCH,  on $Paused,   using { pause_sleep($_, 0.2) };
  transition STATUS, to TOUCH,  on $Running,  using { $_->{finish} = $_->{call}->() ? 0 : 1; $_ };

  # Touch pid file
  transition TOUCH,  to STATUS, using { $_->{pidfile}->touch; $_ };

  # PidFile lock
  transition STATUS, to LOCK,   on $Unlocked, using { $_->{lock} = $_->{pidfile}->lock; $_ };
  transition LOCK,   to STATUS, on $Locked,   using { sigtrap($_); $_->{started} = 1; $_ };
  transition LOCK,   to TERM,   on $Unlocked;

  # Term
  transition STOP,   to TERM,   on $Proctor,  using { undef $_->{lock}; undef $SIG{$_} foreach @{$_->{traps}}; $_ };
};

has pidfile_path => (is => 'ro', isa => NonEmptyStr, required => 1);
has stop_path    => (is => 'ro', isa => NonEmptyStr, required => 1);
has pause_path   => (is => 'ro', isa => NonEmptyStr, required => 1);
has traps        => (is => 'ro', isa => SignalList, default => sub {[]});

has pidfile => (
  is  => 'lazy',
  isa => $PidFile,
  handles => {
    read_pid    => 'read_file',
    running_pid => 'running_pid',
    is_running  => 'is_running',
  },
);

has stop_flag => (
  is  => 'lazy',
  isa => $Flag,
  handles => {
    stop       => 'set',
    start      => 'unset',
    is_stopped => 'is_set',
  },
);

has pause_flag => (
  is  => 'lazy',
  isa => $Flag,
  handles => {
    pause     => 'set',
    resume    => 'unset',
    is_paused => 'is_set',
  },
);

sub _build_pidfile    { Proc::tored::PidFile->new(file_path => shift->pidfile_path) }
sub _build_stop_flag  { Proc::tored::Flag->new(touch_file_path => shift->stop_path) }
sub _build_pause_flag { Proc::tored::Flag->new(touch_file_path => shift->pause_path) }

sub clear_flags {
  my $self = shift;
  $self->start;
  $self->resume;
}

sub run {
  my ($self, $code) = @_;

  my $acc = {
    pidfile => $self->pidfile,
    stopped => $self->stop_flag,
    paused  => $self->pause_flag,
    traps   => $self->traps,
    call    => $code,
    lock    => undef,
    started => 0,
    finish  => 0,
    quit    => 0,
  };

  my $service = $FSM->();

  $service->($acc);

  return $acc->{started};
};

1;


lib/Proc/tored/Manager.pm
package Proc::tored::Manager;
# ABSTRACT: OO interface to creating a proctored service

=head1 SYNOPSIS

  my $proctor = Proc::tored::Manager->new(dir => '/tmp', name => 'my-service');

  # Call do_stuff while the service is running or until do_stuff returns false
  $proctor->service(\&do_stuff)
    or die sprintf('process %d is already running this service!', $proctor->running_pid);

  # Signal another process running this service to quit gracefully, throwing an
  # error if it does not self-terminate after 15 seconds.
  if (my $pid = $proctor->stop_wait(15)) {
    die "process $pid is being stubborn!";
  }

=head1 DESCRIPTION

Objective interface for creating and managing a proctored service.

=cut

use strict;
use warnings;
use Moo;
use Carp;
use Fcntl qw(:flock :seek :DEFAULT);
use Path::Tiny qw(path);
use Time::HiRes qw(sleep);
use Try::Tiny;
use Types::Standard -all;
use Proc::tored::Flag;
use Proc::tored::Machine;
use Proc::tored::PidFile;
use Proc::tored::Types -types;

=head1 METHODS

=head2 new

Creates a new service object, which can be used to run the service and/or
signal another process to quit. The pid file is not created or accessed by this
method.

=over

=item name

The name of the service. Services created with an identical L</name> and
L</dir> will use the same pid file and share flags.

=item dir

A valid run directory (C</var/run> is a common choice). The path must be
writable.

=item pid_file

Unless manually specified, the pid file's path is L</dir>/L</name>.pid.

=item stop_file

Unless manually specified, the stop file's path is L</dir>/L</name>.stopped.

=item pause_file

Unless manually specified, the pause file's path is L</dir>/L</name>.paused.

=item trap_signals

An optional array of signals (suitable for use in C<%SIG>) allowed to end the
L</service> loop. Unless specified, no signal handlers are installed.

=back

=cut

has name         => (is => 'ro', isa => NonEmptyStr, required => 1);
has dir          => (is => 'ro', isa => Dir, required => 1);
has pid_file     => (is => 'lazy', isa => NonEmptyStr);
has stop_file    => (is => 'lazy', isa => NonEmptyStr);
has pause_file   => (is => 'lazy', isa => NonEmptyStr);
has trap_signals => (is => 'ro', isa => SignalList, default => sub {[]});

sub _build_pid_file {
  my $self = shift;
  my $file = path($self->dir)->child($self->name . '.pid');
  return "$file";
}

sub _build_stop_file {
  my $self = shift;
  my $file = path($self->dir)->child($self->name . '.stopped');
  return "$file";
}

sub _build_pause_file {
  my $self = shift;
  my $file = path($self->dir)->child($self->name . '.paused');
  return "$file";
}

has machine => (
  is  => 'lazy',
  isa => InstanceOf['Proc::tored::Machine'],
  handles => [qw(
    clear_flags
    stop start is_stopped
    pause resume is_paused
    read_pid running_pid is_running
  )],
);

sub _build_machine {
  my $self = shift;
  Proc::tored::Machine->new(
    pidfile_path => $self->pid_file,
    stop_path    => $self->stop_file,
    pause_path   => $self->pause_file,
    traps        => $self->trap_signals,
  );
}

=head1 METHODS

=head2 read_pid

Returns the pid identified in the pid file. Returns 0 if the pid file does
not exist or is empty.

=head2 running_pid

Returns the pid of an already-running process or 0 if the pid file does not
exist, is empty, or the process identified by the pid does not exist or is not
visible.

=head2 stop

=head2 start

=head2 is_stopped

Controls and inspects the "stopped" flag. While stopped, the L</service> loop
will refuse to run.

=head2 pause

=head2 resume

=head2 is_paused

Controls and inspects the "paused" flag. While paused, the L</service> loop
will continue to run but will not execute the code block passed in.

=head2 clear_flags

Clears both the "stopped" and "paused" flags.

=head2 is_running

Returns true if the current process is the active, running process.

=head2 stop_wait

Sets the "stopped" flag and blocks until the L<running_pid> exits or the
C<$timeout> is reached.

  $service->stop_wait(30); # stop and block for up to 30 seconds

=cut

sub stop_wait {
  my ($self, $timeout, $sleep) = @_;
  $sleep ||= 0.2;

  $self->stop;
  return if $self->is_running;

  my $pid = $self->running_pid || return 0;

  while (kill(0, $pid) && $timeout > 0) {
    sleep $sleep;
    $timeout -= $sleep;
  }

  !kill(0, $pid);
}

=head2 service

Accepts a code ref which will be called repeatedly until it returns false or
the "stopped" flag is set. If the "paused" flag is set, will continue to rune
but will not execute the code block until the "paused" flag has been cleared.

Example using a pool of forked workers, an imaginary task queue, and a
secondary condition that decides whether to stop running.

  $proctor->service(sub {
    # Wait for an available worker, but with a timeout
    my $worker = $worker_pool->next_available(0.1);

    if ($worker) {
      # Pull next task from the queue with a 0.1s timeout
      my $task = poll_queue_with_timeout(0.1);

      if ($task) {
        $worker->assign($task);
      }
    }

    return if service_should_stop();
    return 1;
  });

=cut

sub service {
  my ($self, $code) = @_;
  $self->machine->run($code);
}

1;


lib/Proc/tored/PidFile.pm
package Proc::tored::PidFile;
# ABSTRACT: Manage a service using a pid file

=head1 SYNOPSIS

  use Proc::tored::PidFile;

  my $pidfile = Proc::tored::PidFile->new(file_path => $pid_file_path);

  if (my $lock = $pidfile->lock) {
    run_service;
  }
  else {
    die "service is already running under process id "
      . $pidfile->running_pid;
  }

=head1 DESCRIPTION

Allows the use of a pid file to manage a running service.

=cut

use warnings;
use strict;
use Moo;
use Carp;
use Guard qw(guard);
use Path::Tiny qw(path);
use Try::Tiny;
use Types::Standard -types;
use Proc::tored::LockFile;

has file_path => (is => 'ro', isa => Str, required => 1);

has file => (is => 'lazy', isa => InstanceOf['Path::Tiny'], handles => ['touch']);
sub _build_file { path(shift->file_path) }

has lockfile => (is => 'lazy', isa => InstanceOf['Proc::tored::LockFile']);
sub _build_lockfile { Proc::tored::LockFile->new(file_path => shift->file_path . '.lock') }

has mypid => (is => 'ro', isa => Int, default => sub { $$ }, init_arg => undef);

=head1 METHODS

=head2 is_running

Returns true if the pid indicated by the pid file is the current process.

=cut

sub is_running {
  my $self = shift;
  $self->running_pid == $$;
}

=head2 running_pid

Returns true if the pid indicated by the pid file is an active, running
process. This is determined by attempting to signal the process using C<kill(0,
$pid)>.

=cut

sub running_pid {
  my $self = shift;
  my $pid = $self->read_file;
  return 0 unless $pid;
  return $pid if kill 0, $pid;
  return 0;
}

=head2 read_file

Returns the pid stored in the pid file or 0 if the pid file does not exist or
is empty.

=cut

sub read_file {
  my $self = shift;
  return 0 unless $self->file->is_file;
  my ($line) = $self->file->lines({count => 1, chomp => 1}) or return 0;
  my ($pid) = $line =~ /^(\d+)$/;
  return $pid || 0;
}

=head2 write_file

Writes the current process id to the pid file. Returns true on success, false
if the pid file exists and contains a running process id or if unable to
atomically write the pid file out.

=cut

sub write_file {
  my $self = shift;
  my $lock = $self->write_lock or return 0;
  return 0 if $self->running_pid;
  $self->file->spew("$$\n");
  return 1;
}

=head2 clear_file

Truncates the pid file and then unlinks it.

=cut

sub clear_file {
  my $self = shift;
  my $lock = $self->write_lock or return;
  return unless $self->is_running;
  return unless $self->file->exists;
  $self->file->append({truncate => 1});
  try { $self->file->remove }
  catch { warn "error unlinking pid file: $_" }
}

=head2 lock

Attempts to write the current process id to the pid file and returns a L<Guard>
that will truncate and unlink the pid file if it goes out of scope.

  {
    my $lock = $pidfile->lock;
    run_service;
  }

  # $lock goes out of scope and pid file is truncated and unlinked

=cut

sub lock {
  my $self = shift;
  return guard { $self->clear_file } if $self->write_file;
  return;
}

#-------------------------------------------------------------------------------
# Creates a .lock file based on $self->pid_file. While the file exists, the
# lock is considered to be held. Returns a Guard that removes the file.
#-------------------------------------------------------------------------------
sub write_lock {
  my $self = shift;
  return unless $$ eq $self->mypid;
  return unless $self->lockfile;
  return $self->lockfile->lock;
}

1;


lib/Proc/tored/LockFile.pm
package Proc::tored::LockFile;
# ABSTRACT: Guard actions with atomic writes

=head1 SYNOPSIS

  use Proc::tored::LockFile;

  my $lockfile = Proc::tored::LockFile->new(file_path => '/path/to/something.lock');

  if (my $lock = $lockfile->lock) {
    ...
  }

=cut

use warnings;
use strict;
use Moo;
use Carp;
use Guard qw(guard);
use Path::Tiny qw(path);
use Try::Tiny;
use Types::Standard -types;

=head1 ATTRIBUTES

=head2 file_path

Path where lock file should be created.

=cut

has file_path => (
  is  => 'ro',
  isa => Str,
  required => 1,
);

has file => (
  is  => 'lazy',
  isa => InstanceOf['Path::Tiny'],
  handles => [qw(exists)],
);

sub _build_file { path(shift->file_path) }

=head1 METHODS

=head2 lock

Attempts to lock the guarded resource by created a new file at L</file_path>.
If the file could not be created because it already exists (using
C<O_CREAT|O_EXCL>), the lock attempt fails and undef is returned. If the lock
is successfully acquired, a L<Guard> object is returned that will unlink the
lock file as it falls out of scope.

=cut

sub lock {
  my $self = shift;

  # Existing lock file means another process came in ahead
  return if $self->exists;

  my $locked = try {
      $self->file->filehandle({exclusive => 1}, '>');
    }
    catch {
      # Rethrow if error was something other than the file already existing.
      # Assume any 'sysopen' error matching 'File exists' is an indication
      # of that.
      die $_
        unless $_->{op} eq 'sysopen' && $_->{err} =~ /File exists/i
            || $self->exists;
    };

  return unless $locked;

  return guard {
    try { $self->exists && $self->file->remove }
    catch { carp "unable to remove lock file: $_" }
  };
}

1;


lib/Proc/tored/Role/Running.pm
package Proc::tored::Role::Running;
# ABSTRACT: add running state and signal handling to services

use strict;
use warnings;
use Moo::Role;
use Carp;
use Path::Tiny 'path';
use Types::Standard -types;

=head1 SYNOPSIS

  package Some::Thing;
  use Moo;

  with 'Proc::tored::Running';

  sub run {
    my $self = shift;

    $self->start;

    while ($self->is_running) {
      do_stuff(...);
    }
  }

=head1 DESCRIPTION

Classes consuming this role are provided with controls to L</start> and
L</stop> voluntarily through the use of a touch file to signal the process to
self-terminate.

=head1 ATTRIBUTES

=head2 term_file

Specifies a file path that will be used as a touch file to signal the process
to self-terminate.

=cut

has term_file => (
  is  => 'ro',
  isa => Maybe[Str],
  required => 1,
);

has _path => (
  is  => 'lazy',
  isa => InstanceOf['Path::Tiny'],
);

sub _build__path { path(shift->term_file) }

has is_started => (
  is  => 'ro',
  isa => Bool,
  default => 0,
);

=head1 METHODS

=head2 is_running

Returns true while the service is running in the current process.

=cut

sub is_running {
  my $self = shift;

  if ($self->_path->exists) {
    $self->{is_started} = 0;
    $self->_path->remove;
  }

  return $self->is_started;
}

=head2 start

Flags the current process as I<running>. After calling this method,
L</is_running> will return true.

=cut

sub start {
  my $self = shift;
  return if $self->is_running;
  $self->_path->remove;
  $self->{is_started} = 1;
  return 1;
}

=head2 stop

Flags the current process as I<not running> and restores any previously
configured signal handlers. Once this method has been called, L</is_running>
will return false.

=cut

sub stop { $_[0]->{is_started} = 0; return 1 }

=head2 signal

Signals the process to stop running by creating a touch file.

=cut

sub signal { $_[0]->_path->touch->assert(sub { $_->exists }) }

1;


lib/Proc/tored.pm
package Proc::tored;
# ABSTRACT: Service management using a pid file and touch files

use strict;
use warnings;
require Exporter;
require Proc::tored::Manager;

=head1 SYNOPSIS

  use Proc::tored;
  use Getopt::Long;

  my %opt = (
    pause  => 0,
    resume => 0,
    stop   => 0,
    zap    => 0,
    start  => 0,
    run    => 0,
  );

  GetOptions(
    'pause'  => \$opt{pause},
    'resume' => \$opt{resume},
    'stop'   => \$opt{stop},
    'zap'    => \$opt{zap},
    'start'  => \$opt{start},
    'run'    => \$opt{run},
  );

  my $service = service 'stuff-doer', in '/var/run';
  my $pid = running $service;

  print "Running service found with pid $pid\n"
    if $pid;

  if ($opt{pause}) {
    # Set the paused flag, causing a running service to block until unset
    pause $service;
  }
  elsif ($opt{resume}) {
    # Unset the paused flag, unblocking any running service
    resume $service;
  }
  elsif ($opt{stop}) {
    # Set the stopped state, preventing new processes from starting the service
    # and causing running processes to self-terminate
    stop $service;
  }
  elsif ($opt{zap}) {
    # Terminate a running process, timing out after 15s
    zap $service, 15
      or die "stuff_doer $pid is being stubborn";
  }
  elsif ($opt{start}) {
    # Allow the service to start running again
    start $service;
  }

  if ($opt{run}) {
    # Run service (if not stopped)
    run { do_stuff() } $service;
  }

=head1 DESCRIPTION

A C<Proc::tored> service is voluntarily managed by a pid file and touch files.

C<Proc::tored> services are specified with a name and a path. Any services
created using the same name and path are considered the same service and will
be aware of other processes via their L</PID FILE> and respect service control
L</FLAGS>.

=head1 EXPORTED SUBROUTINES

All routines are exported by default.

=head2 service

=head2 in

=head2 trap

A proctored service is defined using the C<service> function. The name given to
the service is used in the naming of various files used to control the service
(e.g., pid file and touch files). The C<in> function is used to specify the
local directory where these files will be created and looked for. Signals may
be trapped using C<trap> on non-C<MSWin32> systems.

  my $service = service 'name-of-service', in '/var/run', trap ['TERM', 'INT'];

=head2 pid

Reads and returns the contents of the pid file. Does not check to determine
whether the pid is valid. Returns 0 if the pid file is not found or is empty.

  printf "service may be running under pid %d", pid $service;

=head2 running

Reads and returns the contents of the pid file after checking that the process
identified still exists. Essentially the same as C<kill(0, pid $service)>.
Returns 0 if the pid is not found or cannot be signalled.

  if (my $pid = running $service) {
    warn "service is already running under pid $pid";
  }

=head2 run

Begins the service in the current process. The service, specified as a code
block, will be called until it returns false or the L</stopped> flag is set.

If the L</paused> flag is set, the loop will continue to run without executing
the code block until it has been L</resume>d. If the L</paused> flag is set at
the time C<run> is called, the loop will start but will not begin executing the
code block until the flag is cleared.

If the L</stopped> flag is set, the loop will terminate at the completion of
the current iteration. If the L</stopped> flag is set at the time C<run> is
called, C<run> will return false immediately. The behavior under L</stopped>
takes priority over that of L</paused>.

  my $started = time;
  my $max_run_time = 300;

  run {
    if (time - $started > $max_run_time) {
      warn "Max run time ($max_run_time seconds) exceeded\n";
      warn "  -shutting down\n";
      return 0;
    }
    else {
      do_some_work();
    }

    return 1;
  } $service;

=head2 zap

Sets the "stopped" flag (see L</stop>), then blocks until a running service
exits. Returns immediately (after setting the "stopped" flag) if the
L</running> service is the current process.

  sub stop_service {
    if (my $pid = running $service) {
      print "Attempting to stop running service running under process $pid\n";

      if (zap $pid, 30) {
        print "  -Service shut down\n";
        return 1;
      }
      else {
        print "  -Timed out before service shut down\n";
        return 0;
      }
    }
  }

=head2 stop

=head2 start

=head2 stopped

Controls and inspects the "stopped" flag for the service.

  # Stop a running service
  if (!stopped $service && running $service) {
    stop $service;
  }

  do_work_while_stopped();

  # Allow service to start
  # Note that this does not launch the service process. It simply clears the
  # "stopped" flag that would have prevented it from running again.
  start $service;

=head2 pause

=head2 resume

=head2 paused

Controls and inspects the "paused" flag for the service. In general, this
should never be done inside the L</run> loop (see the warning in L</Pause and
resume>).

  # Pause a running service
  # Note that the running service will not exit. Instead, it will stop
  # executing its main loop until the "paused" flag is cleared.
  if (!paused $service && running $service) {
    pause $service;
  }

  do_work_while_paused();

  # Allow service to resume execution
  resume $service;

=head1 PID FILE

A pid file is used to identify a running service. While the service is running,
barring any outside interference, the pid will contain the pid of the running
process and a newline. After the service process stops, the pid file will be
truncated. The file will be located in the directory specified by L</in>. Its
name is the concatenation of the service name and ".pid".

=head1 FLAGS

Service control flags are persistent until unset. Their status is determined by
the existence of a touch file.

=head2 stopped

A touch file indicating that a running service should self-terminate and that
new processes should not start is created with L</stop> and removed with
L</start>. It is located in the directory specified by L</in>. Its name is the
concatenation of the service name and ".stopped".

=head2 paused

A touch file indicating that a running service should temporarily stop
executing and that new processes should start but not yet execute any service
code is created with L</pause> and removed with L</resume>. It is located in
the directory specified by L</in>. Its name is the concatenation of the service
name and ".paused".

=head1 BUGS AND LIMITATIONS

=head2 Pause and resume

When a service is L</paused>, the code block passed to L</run> is no longer
executed until I<something> calls L</resume>. This can lead to deadlock if
there is no external actor willing to L</resume> the service.

For example, this service will never resume:

  run {
    my $empty = out_of_tasks();

    if ($empty) {
      pause $service;
    }
    elsif (paused $service && !$empty) {
      # This line is never reached because this code block is no longer
      # executed after being paused above.
      resume $service;
    }

    do_next_task();
    return 1;
  } $service;

In most cases, pausing and resuming a service should be handled from outside of
L</run>.

=cut

use parent 'Exporter';

our @EXPORT = qw(
  service
  in
  trap

  pid
  running
  zap
  run

  stop
  start
  stopped

  pause
  resume
  paused
);

sub service ($%)  { Proc::tored::Manager->new(name => shift, @_) }
sub in      ($;@) { dir => shift, @_ }
sub trap    ($)   { trap_signals => shift, @_ }

sub pid     ($)   { $_[0]->read_pid }
sub running ($)   { $_[0]->running_pid }
sub zap     ($;@) { shift->stop_wait(@_) }
sub run     (&$)  { $_[1]->service($_[0]) }

sub stop    ($)   { $_[0]->stop }
sub start   ($)   { $_[0]->start }
sub stopped ($)   { $_[0]->is_stopped }

sub pause   ($)   { $_[0]->pause }
sub resume  ($)   { $_[0]->resume }
sub paused  ($)   { $_[0]->is_paused }

1;
